src/app/progress/progress.tsx

```diff
@@ -0,0 +1,222 @@
+import { sum } from '../util';
+import { t } from 'i18next';
+import { ViewPager, Frame, Track, View } from 'react-view-pager';
+import { isPhonePortraitStream, isPhonePortrait } from '../mediaQueries';
+import './progress.scss';
+
+/* Label isn't used, but it helps us understand what each one is */
+const progressionMeta = {
+  611314723: { label: "Vanguard", order: 1 },
+  3231773039: { label: "Vanguard Research", order: 2 },
+  697030790: { label: "Crucible", order: 3 },
+  1021210278: { label: "Gunsmith", order: 4 },
+
+  4235119312: { label: "EDZ Deadzone Scout", order: 5 },
+  4196149087: { label: "Titan Field Commander", order: 6 },
+  1660497607: { label: "Nessus AI", order: 7 },
+  828982195: { label: "Io Researcher", order: 8 },
+  2677528157: { label: "Follower of Osiris", order: 9 },
+
+  2105209711: { label: "New Monarchy", order: 10 },
+  1714509342: { label: "Future War Cult", order: 11 },
+  3398051042: { label: "Dead Orbit", order: 12 },
+  3468066401: { label: "The Nine", order: 13 },
+  1761642340: { label: "Iron Banner", order: 14 },
+
+  1482334108: { label: "Leviathan", order: 15 }
+};
+
+interface Props {
+  ProgressService: ProgressService;
+  $scope: IScope;
+  account: DestinyAccount;
+  dimSettingsService;
+}
+
+interface State {
+  progress?: ProgressProfile;
+  characterOrder: string;
+  isPhonePortrait: boolean;
+}
+
+export class Progress extends React.Component<Props, State> {
+  subscription: Subscription;
+  mediaQuerySubscription: Subscription;
+
+  constructor(props: Props) {
+    super(props);
+    this.state = {
+      characterOrder: this.props.dimSettingsService.characterOrder,
+      isPhonePortrait: isPhonePortrait()
+    };
+  }
+
+  componentDidMount() {
+    this.subscription = this.props.ProgressService.getProgressStream(this.props.account).subscribe((progress) => {
+      this.setState({ progress });
+    });
+
+    this.mediaQuerySubscription = isPhonePortraitStream().subscribe((isPhonePortrait: boolean) => {
+      if (isPhonePortrait != this.state.isPhonePortrait) {
+        this.setState({ isPhonePortrait });
+      }
+    })
+
+    this.props.$scope.$on('dim-refresh', () => {
+      this.props.ProgressService.reloadProgress();
+    });
+
+    this.props.$scope.$watch(() => this.props.dimSettingsService.characterOrder, (newValue: string) => {
+      if (newValue != this.state.characterOrder) {
+        this.setState({ characterOrder: newValue });
+      }
+    });
+  }
+
+  componentWillUnmount() {
+    if (this.subscription) {
+      this.subscription.unsubscribe();
+    }
+  }
+
+  render() {
+    if (!this.state.progress) {
+      return <div className="progress dim-page">Loading...</div>;
+    }
+
+    const { defs, profileInfo, lastPlayedDate } = this.state.progress;
+
+    const characters = sortCharacters(Object.values(profileInfo.characters.data), this.state.characterOrder);
+
+    // TODO: sorting these is a mystery
+    function milestonesForCharacter(character): IDestinyMilestone[] {
```

Comment:
as `render()` is called every quite frequently, do these functions need to be redeclared every on every render? It's a bit of an anti pattern :/

Perhaps they could be methods on the React Component, or just regular functions (in a seperate file?) that are called with the right args.

Reply:
I'm not sure it's an antipattern to declare nested functions to help out - any decent compiler will turn them into functions that take in-scope variables as parameters, and reuse them. I might move them out if they're something I want to share, or if they can be as nice to use as external functions or methods, but I always like using functions to help with this sort of work, and so far there's no need for them to be available outside the scope of this function.